### PR TITLE
Uncertainty: Quantile Regression V2

### PR DESCRIPTION
**Pull request by Kevin-Chen0**
This is the latest attempt to implement QR as the first uncertainty estimation method for NeuralProphet, rebased on the latest master branch.

Original work is done by @HansikaPH. Link [here](https://github.com/ourownstory/neural_prophet/pull/255).

### CODE REVIEW

File: neuralprophet/configure.py [hunk 1]
```diff
@@ -96,6 +98,25 @@ class Train:
     loss_func_name: str = field(init=False)
 
     def __post_init__(self):
+        # convert to list
+        if not isinstance(self.quantiles, list):
```
Comment (ourownstory) [inline neuralprophet/configure.py:102]: maybe we can just assert that it is a list, instead of wrapping it in a list, to avoid issues with None cases etc?

File: neuralprophet/time_net.py [hunk 2]
```diff
@@ -102,14 +105,19 @@ def __init__(
         # General
         self.n_forecasts = n_forecasts
 
+        # Quantiles
+        self.quantiles = quantiles
+
         # Bias
-        self.bias = new_param(dims=[1])
+        # dimensions - [no. of quantiles, bias shape]
```
Comment (ourownstory) [inline neuralprophet/time_net.py:112]: maybe we should have the quantiles always as the last dimension?

File: neuralprophet/forecaster.py [hunk 3]
```diff
@@ -2602,13 +2648,21 @@ def _reshape_raw_predictions_to_forecst_df(self, df, predicted, components):
...
-        for forecast_lag in range(1, self.n_forecasts + 1):
-            forecast = predicted[:, forecast_lag - 1]
-            pad_before = self.max_lags + forecast_lag - 1
-            pad_after = self.n_forecasts - forecast_lag
-            yhat = np.concatenate(([None] * pad_before, forecast, [None] * pad_after))
-            df_forecast["yhat{}".format(forecast_lag)] = yhat
-            df_forecast["residual{}".format(forecast_lag)] = yhat - df_forecast["y"]
+        for j in range(len(self.config_train.quantiles)):
+            for forecast_lag in range(1, self.n_forecasts + 1):
+                ### ???
```
[the remark this answers is not in the document]
Comment (Kevin-Chen0) [reply neuralprophet/forecaster.py:2653]: no, will remove

File: neuralprophet/custom_loss_metrics.py [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+        Returns:
+            pinball loss (float)
+        """
+        target = target.repeat(1, len(self.quantiles), 1)  # increase the quantile dimension of the targets
+        differences = target - outputs
+        base_losses = self.loss_func(outputs, target)  # dimensions - [n_batch, no. of quantiles, n_forecasts]
+        positive_losses = torch.tensor(self.quantiles).unsqueeze(dim=-1).unsqueeze(dim=0) * base_losses
+        negative_losses = (1 - torch.tensor(self.quantiles).unsqueeze(dim=-1).unsqueeze(dim=0)) * base_losses
+        pinball_losses = torch.where(differences >= 0, positive_losses, negative_losses)
+        # multiplier = torch.ones(size=(1, len(self.quantiles), 1))
```
[the remark this answers is not in the document]
Comment (Kevin-Chen0) [reply neuralprophet/custom_loss_metrics.py:35]: I can double the median weight, but the median quantile loss will be double the weight of the other quantile loss, regardless of how many quantiles there are. Alternatively, I can double the losses for all of the quantiles so that they are "scaled" to the base loss. For example, **2*(0.1 or 0.9)*bass_loss** for a 90th quantile depending on positive or negative loss. All of the quantile losses will averaged out.

File: neuralprophet/utils.py [hunk 5]
```diff
@@ -47,7 +47,7 @@ def reg_func_trend(weights, threshold=None):
     abs_weights = torch.abs(weights)
     if threshold is not None and not math.isclose(threshold, 0):
         abs_weights = torch.clamp(abs_weights - threshold, min=0.0)
-    reg = torch.sum(abs_weights).squeeze()
+    reg = torch.mean(torch.sum(abs_weights, dim=1)).squeeze()
```
[the remark this answers is not in the document]
Comment (Kevin-Chen0) [reply neuralprophet/utils.py:50]: Not sure, this is Hansika's code. Is there a test code/function that runs this?

File: neuralprophet/custom_loss_metrics.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (ourownstory) [reply neuralprophet/custom_loss_metrics.py:35]: based on our conversation: let's double the median and sum over all quantiles at the end

[the remark this answers is not in the document]
Comment (Kevin-Chen0) [reply neuralprophet/custom_loss_metrics.py]: done

File: neuralprophet/time_net.py (hunk 2 above)
Comment (Kevin-Chen0) [reply neuralprophet/time_net.py]: Per our discussion, quantiles will only do after batch and no. forecasts.

File: neuralprophet/configure.py (hunk 1 above)
Comment (Kevin-Chen0) [reply neuralprophet/configure.py:102]: Done

File: neuralprophet/time_net.py [hunk 6]
```diff
@@ -304,10 +327,65 @@ def get_reg_weights(self, name):
...
+            out[:, 0, :] = diffs[:, 0, :]  # set the median where 0 is the median quantile index
+
+            if n_upper_quantiles > 0:  # check if upper quantiles exist
+                upper_quantile_diffs = diffs[:, quantiles_divider_index:, :]
+                if predict_mode:  # check for quantile crossing and correct them in predict mode
+                    upper_quantile_diffs[:, 0, :] = torch.max(torch.tensor(0), upper_quantile_diffs[:, 0, :])
+                    for i in range(n_upper_quantiles - 1):
+                        next_diff = upper_quantile_diffs[:, i + 1, :]
+                        diff = upper_quantile_diffs[:, i, :]
+                        upper_quantile_diffs[:, i + 1, :] = torch.max(next_diff, diff)
```
[the remark this answers is not in the document]
Comment (Kevin-Chen0) [reply neuralprophet/time_net.py]: I think we should ignore the QR crossing for now and do another PR later if we want to fix it with a proper solution.

File: neuralprophet/forecaster.py [hunk 7]
```diff
@@ -1481,6 +1520,7 @@ def plot_components(self, fcst, df_name=None, figsize=None, residuals=False):
         return plot_components(
             m=self,
             fcst=fcst,
+            quantile=0.5,  # ???
```
[the remark this answers is not in the document]
Comment (Kevin-Chen0) [reply neuralprophet/forecaster.py]: the `plot()` does plot all of the quantiles but `plot_components()` only plots the median atm. I've changed it to `quantile=self.config_train.quantiles[0]`.

File: neuralprophet/forecaster.py [hunk 8]
```diff
@@ -1519,6 +1559,7 @@ def plot_parameters(self, weekly_start=0, yearly_start=0, figsize=None, df_name=
         """
         return plot_parameters(
             m=self,
+            quantile=0.5,  # ???
```
[the remark this answers is not in the document]
Comment (Kevin-Chen0) [reply neuralprophet/forecaster.py]: I've changed it to `quantile=self.config_train.quantiles[0]`, which is also the median.

File: neuralprophet/utils.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (Kevin-Chen0) [reply neuralprophet/utils.py:50]: So here, abs_weights dim is `[no. quantiles, no. trend delta]`. So it sums up the trend delta weights (as done before), then averages the quantiles.

File: neuralprophet/time_net.py [hunk 9]
```diff
@@ -559,7 +649,19 @@ def forward(self, inputs):
                 )
 
         trend = self.trend(t=inputs["time"])
-        out = trend + additive_components + trend.detach() * multiplicative_components
+
+        # 0 is the median quantile index
+        # all multiplicative components are multiplied by the median quantile trend
```
[the remark this answers is not in the document]
Comment (Kevin-Chen0) [reply neuralprophet/time_net.py:668]: Multiplying by the median quantile trend is also from Hansika's code. Since I'm not sure what is the benefit of doing that instead of multiplying by respective quantiles (as you said), I have reverted it back to the original:

`trend + additive_components + trend.detach() * multiplicative_components
`
Nevertheless, I still kept that code (in commented form) just in case.

**Comment by ourownstory** [comment]
Notes for future PR:
* fix Quantiles crossing
* enable plotting of parameters of specific quantiles (currently only mean supported)
* add plotting of quantiles to plotly

**Comment by ourownstory** [comment]
To increase test coverage: 
* add a call to `model.predict(...)` in one of the test functions.
* some plotting functions do not get called (when multiple quantiles and no highlight)

**Comment by ourownstory** [comment]
it seems like in `_compute_quantile_forecasts_from_diffs` make sure that `predict_mode=True` when user calls `model.predict()`.
